utils/compilation_unit.mli

```diff
@@ -0,0 +1,144 @@
+   "startup" files nor external libraries.
+*)
+
+[@@@ocaml.warning "+a-9-40-41-42"]
+
+module Name : sig
+  (** The name of a compilation unit without any "-for-pack" prefix. *)
+  type t
+
+  (** Printing, comparison, sets, maps, etc. *)
+  include Identifiable.S with type t := t
+
+  (** [dummy] is a placeholder for units that does not have a valid name, as
+      in the, or during initialisation of the compiler.  It is not a valid
+      identifier and thus cannot be generated through [of_string]. *)
+  val dummy : t
+
+  (** [of_string s] checks the given module name is a valid compilation unit
+      name and generates its representation. *)
+  val of_string : string -> t
+
+  val to_string : t -> string
+end
+
+module Prefix : sig
+  (** A pack name prefix, as specified to "-for-pack".  Such a prefix may
+      be empty. *)
+  type t
+
+  (** Printing, comparison, sets, maps, etc. *)
+  include Identifiable.S with type t := t
+
+  val empty : t
+
+  (** [parse_for_pack p] returns the list of nested packed modules from a
+      "-for-pack" argument. *)
+  val parse_for_pack : string option -> t
+
+  (* CR mshinwell: clarify comment *)
+  (** [extract_prefix id] returns the prefix of an identifier and its basename,
+      as if it was generated with "-for-pack". *)
+  val extract_prefix : string -> t * Name.t
+
+  val to_list_outermost_pack_first : t -> Name.t list
+
+  val to_string : t -> string
+
+  val is_empty : t -> bool
+
+  val first_component_to_name : t -> Name.t option
+end
+
+(** The name of a compilation unit qualified with any "-for-pack" prefix that
+    was specified when the unit was compiled.  For example if compiling foo.ml
+    with "-for-pack Baz.Bar", the corresponding value of type [t] would
+    represent "Baz.Bar.Foo", with its [name] representing "Foo" and its
+    [prefix] representing "Baz.Bar". *)
+type t
+
+(** Printing, comparison, sets, maps, etc. *)
+include Identifiable.S with type t := t
+
+(** Print only the name of the given compilation unit. *)
+val print_name : Format.formatter -> t -> unit
+
+(** Print the full path of the compilation unit, with a dot between each
+    pair of components. *)
+val print_full_path : Format.formatter -> t -> unit
+
+(** Create a compilation unit with the given [name] (which is not encoded or
+    mangled in any way). *)
+val create : ?for_pack_prefix:Prefix.t -> Name.t -> t
+
+(** Create a compilation unit from the given [name] (which is not encoded or
+    mangled in any way). The "-for-pack" of prefix is extracted if there is
+    any. *)
+val of_string : string -> t
+(* CR mshinwell: It's kind of bad that [create name] and [of_string name]
+   do different things w.r.t. the prefix. *)
+
+(** A distinguished compilation unit for initialisation of mutable state. *)
+val dummy : t
+
+(** A distinguished compilation unit for predefined exceptions. *)
+val predef_exn : t
+
+(** The name of the compilation unit, excluding any [for_pack_prefix]. *)
+val name : t -> Name.t
```

Comment:
Perhaps this should be `basename` by analogy to Unix. (This is particularly desirable since a `Compilation_unit.t` is (currently) often called a "name", and asking the basename of a name is perfectly sensible but asking the name of a name is not.)